Display OSD message and countdown if arming is delayed due to beacon

etracer65 opened the pull request:
Provides a clear indication that arming is delayed for cases where DSHOT beacon is active.

Clears the OSD and displays "DISABLING BEACON" and "ARMING IN X.Y" with an active countdown in tenths of a second while arming is delayed due to DSHOT beacon. Once delay period is over the normal "ARMING" message appears.

If the DSHOT beacon is not active then this delay screen is not displayed.

mikeller commented on src/main/io/osd.c:
```diff
@@ -202,6 +202,13 @@ static const uint8_t osdElementDisplayOrder[] = {
 
 PG_REGISTER_WITH_RESET_FN(osdConfig_t, osdConfig, PG_OSD_CONFIG, 3);
 
+static void osdDisplayCenteredMessage(int row, const char *message)
```
I don't think making this into a permanent facility. We should rather convert the existing 'ARMED' text into a facility that can be positioned like all the other elements, and default it to the existing position.

etracer65 replied:
I didn't really change anything in regards to the "ARMED" message other than make it more nicely centered.  Before it was simply hardcoded into position:
```
    displayWrite(osdDisplayPort, 12, 7, "ARMED");
```
So if we want to make the "ARMED" message configurable that can still be done - but should probably be it's own PR and will also require configurator changes.  The bad thing with making it configurable is that in the current scheme the user could choose to not display it.  Not sure that would be a good change.  Currently it clears the display and displays the message for 0.5s on every arming.

mikeller commented on src/main/io/osd.c:
```diff
@@ -1401,6 +1423,13 @@ STATIC_UNIT_TESTED void osdRefresh(timeUs_t currentTimeUs)
 
+#ifdef USE_DSHOT
+    if (isTryingToArm() && !ARMING_FLAG(ARMED)) {
+        osdShowTryingToArm(currentTimeUs);
```
This will suppress all other information for the duration of the delay?
I don't think it is helpful to withhold this information from pilots for this time. Should we rather add this as a warning?

etracer65 replied:
Yes, the screen will blank except for the message and countdown.  Note that during this time we're not armed so there's no useful flight information.  The idea is to give immediate feedback to the arming switch so the pilot doesn't think arming failed.  So it's meant to be "in your face" that it will be arming momentarily.  Changing it to use a warning is risky because the user could choose to not have warnings displayed.

The secondary value is provide indirect pressure on the pilots to change their configuration to avoid having the beacon active at arming.  By improving their configuration they can avoid this delay that they see every time.

mikeller replied:
Not sure about the value of being so prescriptive and unflexible for the 'ARMED' warning. The same effect could be achieved by making it configurable and defaulting it to on - the pilot could disable it, but that would have to be a conscious decision. Plus they could move it to wherever they like.

But I agree that this is something for after 3.4.

Re the Dshot beeper warning, I don't see why this should be so much more prominent than the other warnings, including all of the other arming disabled reasons.

etracer65 replied:
Well I did actually initially think about making it a warning.  But because of:
1. User could disable warnings
2. Warning size is only 11 characters
3. Not really a "arming disabled" but "arming delayed" message

Item 1 was the big issue.  For item 2 I would need to increase a few bytes to maintain the "ARMING In x.y" countdown feedback.  But this wouldn't provide the opportunity to display why the delay was happening like I do now with showing "DISABLING BEACON".  Item 3 is just semantics.

So item 1 was the biggest deal for me, followed by the loss in information for item 2.  But I'm open to changing it to a warning if we want to accept these issues.  I think it would need to override other possible warnings except for arming disabled reasons - but that will automatically happen becasue `isTryingToArm()` will never be true in these circumstances.

Also as you may have seen I opened an issue and called out Felix and Stefan to see if we can't fix this problem in the ESC firmware going forward.  Hopefully that'll be our long-term solution and we can eliminate this arming delayed code completely.

mikeller replied:
@etracer65:

Re 1.: The same applies to other warnings - but they are on by default, and the expectation is that if the user wilfully disables them they are aware that they won't get them. Plus there is also an up side to this - if we add it to the warnings list that can be disabled, we get another place in the configurator where the users can see the warning, and read an eventual tooltip, so they know what's happening before it's happening (and they will have to go there in order to disable the warning).

Re 2.: 'BEACON ON' and 'ARM IN x.y' will fit, without much loss of information. And the warning list is in order of priority, so we can make sure it gets seen.

ledvinap replied:
did you consider animating ARMED message?
It will be minimal change to UI and user will be notified that arming is pending for some reason ...

etracer65 replied:
@mikeller Ok, I will look into changing into a warning.  The warning field is only defined to be 11 characters so I'm not sure how to display both messages - "BEACON ON" and the "ARM IN x.y" countdown.  I suppose I could display the "BEACON ON" on one row above but there's a slight risk the pilot has the warnings positioned at the top of the screen.

mikeller replied:
Ah, I meant to first flash 'BEACON ON', then count down.

etracer65 commented:
Changed to use the OSD_WARNING element rather than a separate separate screen before arming.  The "BEACON ON" message will appear for the first 0.5 seconds and then "ARM IN x.y" will countdown for the remaining delay.

Also fixed OSD_WARNING element blinking related to battery warnings to only blink the warning element if the battery warnings are actually configured.

Suppress blinking if the arming delay messages are being displayed (makes them difficult to read).

Since I reverted basically all the previous commit and went a different direction I'm not sure having separate commit history is worthwhile.  Let me know if you want to squash.

mikeller commented:
If it's an improvement, it should go into 4.0.

etracer65 commented:
Well maybe I should have labeled it a safety improvement.  The goal was to address the concerns raised in slack about the delay in motors starting to spin when beacon was active.  Leading to confusion and possibly a safety issue of someone reaching for the quad during that time.  Granted it only helps for those with OSD.

Also it was an improvement to a safety feature added late in 3.4 - the beacon arming delay.

mikeller commented:
Good point.
